Feat: #3498 Display episode's published_at

vanduc1102 opened the pull request:
<!--- Prepend PR title with [WIP] if work in progress. Remove when ready for review. -->

<!--- For a timely review/response, please avoid force-pushing additional commits if your PR already received reviews or comments -->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Display episode's published at 

## Related Tickets & Documents

https://github.com/thepracticaldev/dev.to/issues/3498

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

<img width="459" alt="older-than-1-day" src="https://user-images.githubusercontent.com/3014017/66325639-ca781180-e951-11e9-9c69-f3c664e78368.png">
<img width="579" alt="Screenshot at Oct 07 19-19-09" src="https://user-images.githubusercontent.com/3014017/66325640-cb10a800-e951-11e9-9d83-2651b1c36d66.png">
<img width="572" alt="no-published-at" src="https://user-images.githubusercontent.com/3014017/66325642-cba93e80-e951-11e9-9986-47c04133910b.png">


## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [ x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4272) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/thepracticaldev/dev.to?pullRequest=4272) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/thepracticaldev/dev.to?pullRequest=4272) it.</sub>

rhymes commented on app/models/podcast_episode.rb:
```diff
@@ -86,6 +86,14 @@ def published
+  def published_timestamp
+    published_at.utc.iso8601
+  end
+
```
These should go into the `podcast_episode_decorator.rb` file

vanduc1102 replied:
thanks @rhymes for helps, i'm working on the issue.

vanduc1102 replied:
Done! thanks @rhymes

vanduc1102 commented:
> Thanks for the pr, @vanduc1102
> Regarding the failing tests, we are not actually attempting to display unpublished episodes in this case. All episodes are considered published if the podcast itself is published. The tests are failing because having `published_at` is not a strict requirement for a podcast episode record. So, before showing the date we should check if it exists. I also recommend adding a check to the new methods so that they wouldnt fail if we run them on a record that doesn't have `published_at`.
> It would be nice to add a test to check that the date is displayed correctly (to `spec/system/podcasts/*`.
> I will require changes though the pr is WIP to mark it for me and other reviewers.

Thanks @lightalloy, 
I made some changes
- Added check for published_at
- Added testcases
Would you please help to review again?

vanduc1102 commented:
> Hi @vanduc1102, thanks for this PR!
> 
> The tests are failing because they are attempting to render podcast episodes that are not published, which looks like a symptom of something else, because an unpublished episode shouldn't end up in the index page. The controller is actually loading only `available` episodes:
> 
> https://github.com/thepracticaldev/dev.to/blob/580fd11bb9b0e5e6fac0b3691989df3078c77a9a/app/controllers/podcast_episodes_controller.rb#L9-L11

Thanks @rhymes for taking your time to review my PR, 
it is my first PR in RnR, so it kinds of messy. 😭 
please help to review again.

rhymes commented on app/decorators/podcast_episode_decorator.rb:
```diff
@@ -8,4 +8,12 @@ def comments_to_show_count
   end
+
+  def readable_publish_date
+    published_at&.strftime("%b %e")
```
I think this should display the year for older podcasts, see how for example, it is done for the comment:

```ruby
  def readable_publish_date
    if created_at.year == Time.current.year
      created_at.strftime("%b %e")
    else
      created_at.strftime("%b %e '%y")
    end
  end
```

rhymes commented on app/views/articles/_podcast_episodes_feed.html.erb:
```diff
@@ -18,6 +18,12 @@
+        <% end %>
       </h4>
     </div>
   </a>
```
This file doesn't seem to be in use, you can remove it.

rhymes commented on app/views/podcast_episodes/_episodes_feed.html.erb:
```diff
@@ -19,6 +19,12 @@
           <%= episode.podcast.title %>
+          <% if episode.published_at.present? %>
+            ・<time class="published-at" datetime="<%= episode.decorate.published_timestamp %>">
+            <%= episode.decorate.readable_publish_date %>
```
you don't need to call `.decorate`, the collection has been already decorated in the controller

lightalloy replied:
You will also need to decorate the collection inside the `StoriesController#handle_podcast_index`

vanduc1102 replied:
thanks, i'm working on it

lightalloy commented on app/views/articles/_podcast_episodes_feed.html.erb:
```diff
@@ -18,6 +18,12 @@
       <h4>
         <%= episode.podcast.title %>
+        <% if episode.published_at.present? %>
+          ・<time class="published-at" datetime="<%= episode.decorator.published_timestamp %>">
```
`decorator` doesn't seem like a valid method, so rendering this template should fail.
I checked the codebase and found out that it's not used in the codebase. I suppose we should delete it. Please, tell me if you find that it's used and I haven't noticed it.

lightalloy commented on app/views/podcast_episodes/show.html.erb:
```diff
@@ -50,6 +50,11 @@
       <% else %>
         <h1><%= @episode.title %></h1>
       <% end %>
+      <% if @episode.published_at.present? %>
```
Btw, you can check it a bit shorter `@episode.published_at?`

lightalloy commented on app/views/podcast_episodes/show.html.erb:
```diff
@@ -50,6 +50,11 @@
         <h1><%= @episode.title %></h1>
       <% end %>
+      <% if @episode.published_at.present? %>
+        <time class="published-at" datetime="<%= @episode.decorate.published_timestamp %>">
```
I think it's worth decorating in the controller action (`StoriesController#handle_podcast_show`) and not decorate multiple times inside a template.

vanduc1102 commented:
Hi @lightalloy @rhymes 
I addressed the comments

![Screen Shot 2019-10-08 at 5 58 09 PM](https://user-images.githubusercontent.com/3014017/66390677-5e4ee980-e9f5-11e9-821d-66bb48f1efa4.png)

rhymes commented on app/views/articles/_podcast_episodes_feed.html.erb:
```diff
@@ -18,12 +18,6 @@
       </div>
```
I meant to actually remove the file :D I can't find a place in the codebase where this is used

lightalloy replied:
Yeah, the file itself can be deleted :)

vanduc1102 replied:
thanks , deleted

rhymes commented:
@vanduc1102 thanks for the fix and the screenshot. I noticed something weird in the picture, the time ago is displayed only for some episodes in the list

lightalloy commented on app/controllers/stories_controller.rb:
```diff
@@ -141,7 +141,8 @@ def handle_podcast_index
     @article_index = true
     @list_of = "podcast-episodes"
-    @podcast_episodes = @podcast.podcast_episodes.reachable.order("published_at DESC").limit(30)
+    @podcast_episodes = PodcastEpisodeDecorator.decorate_collection(@podcast.podcast_episodes.
```
I believe you can do that with just adding `.decorate` to the query. Haven't noticed that on the first review.

vanduc1102 replied:
would you give me example. i'm new to ruby

lightalloy replied:
I mean `@podcast.podcast_episodes.reachable.order("published_at DESC").limit(30).decorate`

vanduc1102 replied:
Updated, thanks

vanduc1102 commented:
> @vanduc1102 thanks for the fix and the screenshot. I noticed something weird in the picture, the time ago is displayed only for some episodes in the list

Hi @rhymes , i deleted the file `app/views/articles/_podcast_episodes_feed.html.erb`
`the time ago is displayed only for some episodes in the list`, i followed the implementation in https://github.com/thepracticaldev/dev.to/blob/master/app/views/articles/index.html.erb#L121 
which is rendering on dev.to live 
![Screen Shot 2019-10-08 at 7 37 10 PM](https://user-images.githubusercontent.com/3014017/66396062-22bb1c00-ea03-11e9-9855-361017a534b8.png)

rhymes commented:
@vanduc1102 yeah you did right, i'll investigate because that sounds like a bug in both places. But it's out of the scope of this PR anyway!
